transports/quic/src/transport.rs

```diff
@@ -76,48 +78,76 @@ pub enum Error {
     type Dial = Pin<Box<dyn Future<Output = Result<Self::Output, Self::Error>> + Send>>;
 
-    fn listen_on(
-        &mut self,
-        addr: Multiaddr,
-    ) -> Result<Self::Listener, TransportError<Self::Error>> {
-        multiaddr_to_socketaddr(&addr)
-            .ok_or_else(|| TransportError::MultiaddrNotSupported(addr))?;
-        Ok(self.clone())
+    fn listen_on(&mut self, addr: Multiaddr) -> Result<ListenerId, TransportError<Self::Error>> {
+        let socket_addr =
+            multiaddr_to_socketaddr(&addr).ok_or(TransportError::MultiaddrNotSupported(addr))?;
+        let listener_id = ListenerId::new();
+        let listener = Listener::new(listener_id, socket_addr, self.config.clone())
+            .map_err(TransportError::Other)?;
+        self.listeners.push(listener);
+        // Drop reference to dialer endpoint so that the endpoint is dropped once the last
+        // connection that uses it is closed.
+        // New outbound connections will use a bidirectional (listener) endpoint.
+        let _ = self.dialer.take();
+        Ok(listener_id)
     }
 
-    fn address_translation(&self, server: &Multiaddr, observed: &Multiaddr) -> Option<Multiaddr> {
+    fn remove_listener(&mut self, id: ListenerId) -> bool {
+        if let Some(listener) = self.listeners.iter_mut().find(|l| l.listener_id == id) {
+            listener.close(Ok(()));
+            true
+        } else {
+            false
+        }
+    }
+
+    fn address_translation(&self, _server: &Multiaddr, observed: &Multiaddr) -> Option<Multiaddr> {
         Some(observed.clone())
     }
 
     fn dial(&mut self, addr: Multiaddr) -> Result<Self::Dial, TransportError<Self::Error>> {
-        let socket_addr = if let Some(socket_addr) = multiaddr_to_socketaddr(&addr) {
-            if socket_addr.port() == 0 || socket_addr.ip().is_unspecified() {
-                tracing::error!("multiaddr not supported");
-                return Err(TransportError::MultiaddrNotSupported(addr));
-            }
-            socket_addr
-        } else {
+        let socket_addr = multiaddr_to_socketaddr(&addr)
+            .ok_or_else(|| TransportError::MultiaddrNotSupported(addr.clone()))?;
+        if socket_addr.port() == 0 || socket_addr.ip().is_unspecified() {
             tracing::error!("multiaddr not supported");
             return Err(TransportError::MultiaddrNotSupported(addr));
+        }
+        let endpoint = if self.listeners.is_empty() {
+            match self.dialer.clone() {
+                Some(endpoint) => endpoint,
+                None => {
+                    let endpoint =
+                        Endpoint::new_dialer(self.config.clone()).map_err(TransportError::Other)?;
+                    let _ = self.dialer.insert(endpoint.clone());
+                    endpoint
+                }
+            }
+        } else {
+            // Pick a random listener to use for dialing.
+            // TODO: Prefer listeners with same IP version.
```

Comment:
I guess you could re-use the logic used in `libp2p-tcp`?

https://github.com/libp2p/rust-libp2p/blob/2f2b7cbec127a48d869d87ad52d69a4c4c29ddb5/transports/tcp/src/lib.rs#L128-L158